### PR TITLE
feat: named & sharded topic validation and ENR update

### PR DESCRIPTION
**Pull request by SionoiS**
# Description
Process node config for either static sharding or named topics, verify and update the ENR.

# Changes
- Parse topics from config then check their validity according to https://rfc.vac.dev/spec/51/
- Update ENR with sharded topic info.

Step 1 in closing #1776 

### Feedback

All feedback is welcome but more specifically on;

- ~~Had to make some func public. Could be undone in next step when discovery is re-org. Not sure yet. Otherwise maybe we want to change the ergonomics around RelayShard <-> Record.~~
- ~~Not using parsed topics internally seams inelegant. What do?~~
- ~~Could leave the new code where it is and re-org once discovery is changed. Otherwise, where does this new code fit IDK?~~
- ~~Nim "style", tips & tricks. egg. L522 I tried ```return err("Pubsub topic parsing error: " & topic.error)``` I don't understand why it doesn't work? ParsingError has the concat operator non?~~

### CODE REVIEW

File: apps/wakunode2/app.nim [hunk 1]
```diff
@@ -512,6 +512,46 @@ proc setupProtocols(node: WakuNode, conf: WakuNodeConf,
       pubsubTopics = conf.topicsDeprecated.split(" ")
     else:
       pubsubTopics = conf.topics
+
+    # Validate topics and update a node's ENR accordingly
+    let parsed_topics = pubsubTopics.mapIt(NsPubsubTopic.parse(it))
```
Comment (alrevuelta) [inline apps/wakunode2/app.nim:517]: minor, we use `camelCase`.
feel free to have a look
https://hackmd.io/@vac/main/%2F5oIVOuQcSM2WKrnDLUPcpg

File: apps/wakunode2/app.nim [hunk 2]
```diff
@@ -512,6 +512,46 @@ proc setupProtocols(node: WakuNode, conf: WakuNodeConf,
...
+      let key = node.wakuDiscv5.protocol.privKey()
+      
+      let pairs = if (let idxList = relayShard.toIndicesList(); idxList.isOk):
+        [toFieldPair(ShardingIndicesListEnrField, idxList.value)]
+      else:
+        let bitVector = relayShard.toBitVector()
+
+        [toFieldPair(ShardingBitVectorEnrField, bitVector)]
+
+      if record.update(key, pairs).isErr():
```
Comment (alrevuelta) [inline apps/wakunode2/app.nim:552]: wondering if this logic should go in `initNode` or `WakuDiscoveryV5.new`.

`flags = netConfig.wakuFlags.get()` is similar, see [this](https://rfc.vac.dev/spec/31/#waku2-enr-key), so perhaps its a good idea to do all enr field modification in the same place: so far it would be shards + wakuFlags.

File: apps/wakunode2/app.nim [hunk 3]
```diff
@@ -512,6 +512,46 @@ proc setupProtocols(node: WakuNode, conf: WakuNodeConf,
...
+
+    # Validate topics and update a node's ENR accordingly
+    let parsed_topics = pubsubTopics.mapIt(NsPubsubTopic.parse(it))
+
+    var prevTopic: NsPubsubTopic
+    for i, topic in parsed_topics:
+      if topic.isErr():
+        return err("Pubsub topic parsing error")
+      
+      if i == 0:
```
Comment (alrevuelta) [inline apps/wakunode2/app.nim:524]: mind elaborating on this?

File: apps/wakunode2/app.nim (hunk 2 above)
Comment (alrevuelta) [inline apps/wakunode2/app.nim:552]: wondering if there is any other higher level function we can use instead¿

not sure if `withWakuRelayShardingIndicesList` or `withWakuRelaySharding` are meant for this?

perhaps these tests can serve as an example
https://github.com/waku-org/nwaku/blob/0111b88b9b3c13065de6f09844f03c8c0259fcdc/tests/v2/test_waku_enr.nim#L337

**Review by alrevuelta — CHANGES_REQUESTED** [review]
left some comments, thanks!

File: apps/wakunode2/app.nim (hunk 1 above)
Comment (SionoiS) [reply apps/wakunode2/app.nim:517]: Rust habits... come back when I don't pay attention! good spot.

Comment (SionoiS) [reply apps/wakunode2/app.nim:517]: Any Nim tools to automate this?

File: apps/wakunode2/app.nim (hunk 2 above)
Comment (SionoiS) [reply apps/wakunode2/app.nim:552]: The discv5 re-org changes a lot of thing. Notably where the ENR will be created. IDK where it fit yet.

Comment (SionoiS) [reply apps/wakunode2/app.nim:552]: Now that I'm changing where discv5 will sit. I will use the EnrBuilder yes.

File: apps/wakunode2/app.nim [hunk 4]
```diff
@@ -512,6 +512,46 @@ proc setupProtocols(node: WakuNode, conf: WakuNodeConf,
...
+        prevTopic = topic.value
+        continue
+      
+      if prevTopic.kind != topic.value.kind:
```
[the remark this answers is not in the document]
Comment (jm-clius) [reply apps/wakunode2/app.nim:528]: Indeed, this would require iterating twice, but given that it improves readability quite a bit, I think it's worth it.
The 99% use case would be nodes not using any static shards. I'd also suggest extracting most of this logic into a separate unit-testable function, perhaps taking the `parsedTopics` as argument and returning a `Result[RelayShards]`. Then you could do something like:

```
if parsedTopics.anyIt(it.kind == NsPubsubTopicKind.StaticSharding):
  let relayShardsRes = parsedTopics.getRelayShards()
  ## error handling here in case result is err
```

The `getRelayShards` func could then check that `allIt` is StaticSharding and of the same cluster before returning the value.

File: apps/wakunode2/app.nim (hunk 2 above)
Comment (jm-clius) [reply apps/wakunode2/app.nim:552]: Understand there may be some complexity if we don't do the code reorg now, but I'd still prefer (temporarily) extending the constructor of `wakuDiscv5` than exposing the underlying functions.

File: apps/wakunode2/app.nim (hunk 4 above)
[the remark this answers is not in the document]
Comment (SionoiS) [reply apps/wakunode2/app.nim:528]: Is there a way to handle error when using functional style Nim?
```
if parsedTopicsRes.anyIt(it.isErr()):
    return err(it.error)
```
Something like that?

**Comment by SionoiS** [comment]
I still don't get how to handle error functional style but I'm getting better at this!

**Comment by jm-clius** [comment]
I think I've missed a couple of your queries before. :) I hope the following helps:

>  a way to handle error when using functional style Nim

In this case your `Result` has a `string` error type, so you have to stringify (`$`) the error before you can return it. If you had defined an underlying error type or used an existing one, you could in theory define that as the error `Return` type and simply return the underlying error. Quick check shows this being done in e.g. https://github.com/status-im/nim-chronos/blob/0035f4fa6692e85756aa192b4df84c21d3cacacb/chronos/handles.nim#L81-L104

>  I tried return err("Pubsub topic parsing error: " & topic.error) I don't understand why it doesn't work?

Mmm. Perhaps simply needing to stringify: `$topic.error`?

> Not using parsed topics internally seams inelegant. What do?

Not sure parsed topics should have any meaning internally - it's only of use to the application. From the `waku_node` module's perspective a `pubsubTopic` is a non-semantic string.

I see you've made some changes since my last review - I'll still review these ASAP. :)

File: apps/wakunode2/app.nim (hunk 1 above)
Comment (alrevuelta) [reply apps/wakunode2/app.nim]: > Any Nim tools to automate this? (formatting & style)

no that i know 😢

File: waku/v2/waku_discv5.nim [hunk 5]
```diff
@@ -48,6 +48,36 @@ type WakuDiscoveryV5* = ref object
     protocol*: protocol.Protocol
     listening*: bool
 
+func topicsToRelayShards*(topics: seq[string]): Result[Option[RelayShards], string] =
```
Comment (alrevuelta) [inline waku/v2/waku_discv5.nim:51]: perhaps its me but i find this code a bit confusing. there are some special cases i == 0, i > 0, and not sure i like the use of prevTopic.


im for sure missing something so take this as just **pseudocode**. we iterate multiple times but performance here is not an issue, but imho it improves readibility?

```nim
# check errors
if parsedTopicsRes.anyIt(it.isErr()):
    return err(it.error)

# check all are named sharding
if parsedTopicsRes.anyIt(it.get.kind == NsPubsubTopicKind.StaticSharding):
    return error(xxx)

# check all shards belong to the same cluster
if !parsedTopicsRes.allIt(it.get == parsedTopicsRes[0].get.cluster):
   return errors(xxx)

# return
return RelayShards.init(
                 cluster=parsedTopicsRes[0].get.cluster,
                 shards=parsedTopicsRes.mapIt(it.get.shard)

```

**Review by alrevuelta — COMMENTED** [review]
left some comments, thanks!

File: waku/v2/waku_discv5.nim (hunk 5 above)
Comment (jm-clius) [reply waku/v2/waku_discv5.nim:51]: Agree with Alvaro here re readability. And I think it's possible to do only one iteration if you expand your predicate, e.g.:

```
if not parsedTopicsRes.allIt(
  it.isOk() &&
  it.get.kind == NsPubsubTopicKind.StaticSharding &&
  it.get.cluster == parsedTopicsRes[0].get.cluster
  )
  return error(....)
```

If you want to you could even extract the three conditions above into their own predicate function (e.g. `isValid()`), but I think it's readable enough without having to do this. Up to you :)

File: waku/v2/waku_discv5.nim [hunk 6]
```diff
@@ -48,6 +48,36 @@ type WakuDiscoveryV5* = ref object
...
+        return err("use sharded topics within the same cluster.")
+
+    if i > 0:
+      prevTopic = topic.value
+
+  if prevTopic.kind == NsPubsubTopicKind.StaticSharding:
+    let indices = parsedTopicsRes.mapIt(it.value.shard)
+    let relayShards = RelayShards.init(prevTopic.cluster, indices)
+
+    return ok(some(relayShards))
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:77]: Good idea to have the final `return` be a default and not subject to a condition. In other words, the "happy path" for the function returns the final return statement. Every other return is either an error condition or a quick success.

File: tests/v2/test_waku_discv5.nim [hunk 7]
```diff
@@ -280,3 +280,30 @@ procSuite "Waku Discovery v5":
 
     # Cleanup
     await allFutures(node1.stop(), node2.stop(), node3.stop(), node4.stop())
+
+  asyncTest "get relayShards from topics":
+    ## Given
+    let mixedTopics = @["/waku/2/thisisatest", "/waku/2/rs/0/2", "/waku/2/rs/0/8"]
+    let shardedTopics = @["/waku/2/rs/0/2", "/waku/2/rs/0/4", "/waku/2/rs/0/8"]
+    let namedTopics = @["/waku/2/thisisatest", "/waku/2/atestthisis", "/waku/2/isthisatest"]
+    let gibberish = @["aedyttydcb/uioasduyio", "jhdfsjhlsdfjhk/sadjhk", "khfsd/hjfdsgjh/dfs"]
```
Comment (jm-clius) [inline tests/v2/test_waku_discv5.nim:289]: Perhaps also test an empty sequence?

File: waku/v2/waku_discv5.nim [hunk 8]
```diff
@@ -89,6 +126,15 @@ proc new*(T: type WakuDiscoveryV5,
         )
         builder.withWakuCapabilities(flags)
         builder.withMultiaddrs(multiaddrs)
+
+        if relayShardsRes.isOk():
+          let op = relayShardsRes.value
+          if op.isSome():
+            let shards = relayShardsRes.value.get()
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:133]:
```suggestion
            let shards = op.get()
```
perhaps?

File: waku/v2/waku_discv5.nim [hunk 9]
```diff
@@ -89,6 +126,15 @@ proc new*(T: type WakuDiscoveryV5,
...
+            let res = builder.withWakuRelaySharding(shards)
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:134]: I don't necessarily think it's necessary to define `shards` as a separate variable here.
Something like:

```suggestion
            let res = builder.withWakuRelaySharding(op.get())
```

If you want to be more explicit you could rename `op` to e.g. `shardsOp`

File: tests/v2/test_waku_discv5.nim [hunk 10]
```diff
@@ -280,3 +280,30 @@ procSuite "Waku Discovery v5":
...
+
+    let relayShards = RelayShards.init(0, 2, 4, 8)
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply tests/v2/test_waku_discv5.nim:291]: I was confused as well the first thing I tried was indeed ```let relayShards = RelayShards.init(0, @[2, 4, 8])```

File: waku/v2/waku_discv5.nim [hunk 11]
```diff
@@ -77,9 +107,16 @@ proc new*(T: type WakuDiscoveryV5,
...
-          discv5Config: protocol.DiscoveryConfig = protocol.defaultDiscoveryConfig): T {.
+          topics: seq[string],
+          discv5Config: protocol.DiscoveryConfig = protocol.defaultDiscoveryConfig
+          ): T {.
   deprecated: "use the config and record proc variant instead".}=
 
+  let relayShardsRes = topicsToRelayShards(topics)
+
+  if relayShardsRes.isErr():
+    debug "pubsub topic parsing error", reason = relayShardsRes.error
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:118]: May be possible to skip one conditional check later by getting the `shards` option here already.
Something like:
```suggestion
  var relayShardsOp: Option[RelayShard]
  if relayShardsRes.isOk():
    relayShardsOp = relayShardsRes.get()
  else
    debug "pubsub topic parsing error", reason = relayShardsRes.error
    relayShardsOp = none(RelayShard)
```

That way the `record` block below can be a bit cleaner by just checking the option before modifying the builder. Wdyt?

**Review by jm-clius — COMMENTED** [review]
Thanks! I think this is now much better separated. Some of my comments below are nitpicks, so feel free to address as you see fit.

File: waku/v2/waku_discv5.nim (hunk 5 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:51]:
```
if parsedTopicsRes.anyIt(it.isErr()):
    return err(it.error)
```
That is what I wanted to do but seams like you can't use ```it``` outside of scope no?

File: tests/v2/test_waku_discv5.nim (hunk 7 above)
Comment (SionoiS) [reply tests/v2/test_waku_discv5.nim:289]: Good Idea!

File: waku/v2/waku_discv5.nim (hunk 8 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:133]: I might rewrite this it look funky but yes your right!

File: waku/v2/waku_discv5.nim (hunk 5 above)
Comment (jm-clius) [reply waku/v2/waku_discv5.nim:51]: Where do you mean out of scope? The predicate function could be defined inside the scope of the outer `topicsToRelayShards()` func.

The following should work structurally:

```
proc myOuterFunction()...
  
  func isValidPredicate(someVar: VariableType): bool =
    return someBooleanPredicateCondition

  if not myList.allIt(it.isValidPredicate()):
   return error(...)
```

Not great pseudocode, but hope it illustrates what I mean 😅

Of course, in this instance I don't even think the separate predicate `func` is necessary and you could just add the simple conditions within the `allIt()` argument directly.

File: waku/v2/waku_discv5.nim (hunk 11 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:118]: Agree

File: waku/v2/waku_discv5.nim (hunk 5 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:51]: In Rust I would do
```
let result: Result<Vec<_>, Error> = topics.into_iter().map(|item| parse(item)).collect;
let parsed_topics = result?;
```
Can't seams to do the same in Nim... How do you return an Error inside a sequence in a functional style?
```
for res in parsedTopics
   if res.IsErr()
      return err(res.error)
```
That works but it's imperative style and looks ugly :P

**Comment by SionoiS** [comment]
I still don't know if there's a way to do this like in Rust lang
```
let seqOfResults: seq[Result[T, Error]]
```
Into
```
let resultOfSeq: Result[seq[T], Error]
```
Might be something I have talk on the Nim forums

File: waku/v2/waku_discv5.nim [hunk 12]
```diff
@@ -77,9 +95,18 @@ proc new*(T: type WakuDiscoveryV5,
...
+          discv5Config: protocol.DiscoveryConfig = protocol.defaultDiscoveryConfig
+          ): T {.
   deprecated: "use the config and record proc variant instead".}=
 
+  let relayShardsRes = topicsToRelayShards(topics)
+
+  let relayShard: Option[RelayShards] = if relayShardsRes.isErr():
+    debug "pubsub topic parsing error", reason = relayShardsRes.error
+    none(RelayShards)
+  else: relayShardsRes.get()
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:108]: Nitpick:

Perhaps possible to indent the `if` and `else` to be under each other and indented from the `let`.
The following is just a suggestion but may be one way to do it:

```suggestion
  let relayShard =
    if relayShardsRes.isErr():
      debug "pubsub topic parsing error", reason = relayShardsRes.error
      none(RelayShards)
    else: relayShardsRes.get()
```

File: waku/v2/waku_discv5.nim [hunk 13]
```diff
@@ -89,6 +116,12 @@ proc new*(T: type WakuDiscoveryV5,
         )
         builder.withWakuCapabilities(flags)
         builder.withMultiaddrs(multiaddrs)
+
+        if relayShard.isSome():
+          let res = builder.withWakuRelaySharding(relayShard.get())
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:121]: Perhaps a good idea to add `debug` log here to indicate that we're building ENR with relay shard information?

**Review by jm-clius — APPROVED** [review]
Great work! I have some nits below, but it's up to you if you want to address in this or a next PR. I think this incremental PR is ready to be merged, so feel free to unmark as Draft when you agree.

File: waku/v2/waku_discv5.nim (hunk 12 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:108]: I agree, your version is more readable.

File: waku/v2/waku_discv5.nim (hunk 13 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:121]: Agreed

**Comment by SionoiS** [comment]
I asked @arnetheduck about ```seq[Result[T, Error]]``` => ```Result[seq[T], Error]```
He came up with this
```
proc transposeFirstError[T, E](values: openArray[Result[T, E]]): Result[seq[T], E] =
  let errors = mySeq.filterIt(it.isErr())
  if errors.len>0: err(errors[0]) # or some other aggregation strategy
  else: ok(mySeq.mapIt(it.value))
```
Promising!

**Review by Ivansete-status — APPROVED** [review]
That looks awesome from my point of view. Thanks for it!
Just a minor comments :)

**Review by alrevuelta — COMMENTED** [review]
Before approving, can we get a "proof" that it works end to end? For example, configure some shards via cli arguments, run nwaku and paste here the ENR that is generated. Then would like to see in https://enr-viewer.com/ that the fields are the expected ones.

File: waku/v2/waku_discv5.nim [hunk 14]
```diff
@@ -48,6 +48,24 @@ type WakuDiscoveryV5* = ref object
...
+    if res.isErr():
+      return err("failed to parse topic: " & $res.error)
+
+  if parsedTopicsRes.allIt(it.get().kind == NsPubsubTopicKind.NamedSharding):
+    return ok(none(RelayShards))
+
+  if parsedTopicsRes.anyIt(it.get().kind == NsPubsubTopicKind.NamedSharding):
+    return err("use named topics OR sharded ones not both.")
+
+  if parsedTopicsRes.anyIt(it.get().cluster != parsedTopicsRes[0].get().cluster):
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:64]: I believe .allIt returns default true for empty sequence but I guess explicitly checking is more readable.

File: waku/v2/waku_discv5.nim [hunk 15]
```diff
@@ -48,6 +48,24 @@ type WakuDiscoveryV5* = ref object
...
+    return err("use sharded topics within the same cluster.")
+
+  return ok(some(RelayShards.init(parsedTopicsRes[0].get.cluster, parsedTopicsRes.mapIt(it.get().shard))))
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:67]: Annoyingly these are not marked as errors. Good catch!

File: tests/v2/test_waku_discv5.nim [hunk 16]
```diff
@@ -280,3 +279,34 @@ procSuite "Waku Discovery v5":
...
+    let mixedRes = topicsToRelayShards(mixedTopics)
+    let shardedRes = topicsToRelayShards(shardedTopics)
+    let namedRes = topicsToRelayShards(namedTopics)
+    let gibberishRes = topicsToRelayShards(gibberish)
+    let emptyRes = topicsToRelayShards(empty)
+
+    ## Then
+    check:
+      mixedRes.isErr()
+      shardedRes.isOk()
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply tests/v2/test_waku_discv5.nim:304]: I gather that ```check:``` is more for simpler booleans.

**Comment by SionoiS** [comment]
> Before approving, can we get a "proof" that it works end to end? For example, configure some shards via cli arguments, run nwaku and paste here the ENR that is generated. Then would like to see in https://enr-viewer.com/ that the fields are the expected ones.

```enr:-Le4QFjM3fZPD1QR9YdhwD8KXRHfGJMRAFKtSWk6xj-jB5GkTuC5YKZFe9TqDHTFAwCTM_S1OldCiEsiYY0VRwxjvnYBgmlkgnY0gmlwhMCoAZiKbXVsdGlhZGRyc4CCcnOVAAAJAAIAAwAEAAUABgAHAAgACQAKiXNlY3AyNTZrMaED_vsoLc-A90e9uS29Mmd4SKckuakcBnncVHdeztxFfFGDdGNwgupgg3VkcIIjKIV3YWt1MgE```

LGTM

**Review by alrevuelta — APPROVED** [review]
thanks for the ENR, ready to go when CI passes.
